lib/MySQL_Session.cpp

```diff
@@ -348,15 +549,570 @@ int MySQL_Session::handler() {
+							if ((myds->myconn->reusable==true) && myds->myconn->IsActiveTransaction()==false) {
+								retry_conn=true;
+							}
+							myds->destroy_MySQL_Connection_From_Pool();
+							myds->fd=0;
+							if (retry_conn) {
+								myds->DSS=STATE_NOT_INITIALIZED;
+								previous_status.push(PROCESSING_QUERY);
+								NEXT_IMMEDIATE(CONNECTING_SERVER);
+							}
+							return -1;
+						}
+						int myerr=mysql_errno(myconn->mysql);
+						if (myerr > 2000) {
+							bool retry_conn=false;
+							// client error, serious
+							proxy_error("Detected a broken connection during query: %d, %s\n", myerr, mysql_error(myconn->mysql));
+							//if ((myds->myconn->reusable==true) && ((myds->myprot.prot_status & SERVER_STATUS_IN_TRANS)==0)) {
+							if ((myds->myconn->reusable==true) && myds->myconn->IsActiveTransaction()==false) {
+								retry_conn=true;
+							}
+							myds->destroy_MySQL_Connection_From_Pool();
+							myds->fd=0;
+							if (retry_conn) {
+								myds->DSS=STATE_NOT_INITIALIZED;
+								previous_status.push(PROCESSING_QUERY);
+								NEXT_IMMEDIATE(CONNECTING_SERVER);
+							}
+							return -1;
+						} else {
+							proxy_warning("Error during query: %d, %s\n", myerr, mysql_error(myconn->mysql));
+
+							MySQL_Result_to_MySQL_wire(myconn->mysql,myconn->mysql_result,&client_myds->myprot);
+							GloQPro->delete_QP_out(qpo);
+							qpo=NULL;
+							myconn->async_free_result();
+							//myds->DSS=STATE_NOT_INITIALIZED;
+							status=WAITING_CLIENT_DATA;
+							client_myds->DSS=STATE_SLEEP;
+							myds->free_mysql_real_query();
+							//if ((myds->myconn->reusable==true) && ((myds->myprot.prot_status & SERVER_STATUS_IN_TRANS)==0)) {
+							if ((myds->myconn->reusable==true) && myds->myconn->IsActiveTransaction()==false) {
+								myds->DSS=STATE_NOT_INITIALIZED;
+								myds->return_MySQL_Connection_To_Pool();
+							} else {
+								myconn->async_state_machine=ASYNC_IDLE;
+								myds->DSS=STATE_MARIADB_GENERIC;
+							}
+						}
+					} else {
+						// rc==1 , nothing to do for now
+					}
+				}
+
+				goto __exit_DSS__STATE_NOT_INITIALIZED;
+
+
+			}
+			break;
+
+		case CHANGING_USER_SERVER:
+			//fprintf(stderr,"CHANGING_USER\n");
+			assert(mybe->server_myds->myconn);
+			{
+				MySQL_Data_Stream *myds=mybe->server_myds;
+				MySQL_Connection *myconn=myds->myconn;
+				myds->DSS=STATE_MARIADB_QUERY;
+				enum session_status st=status;
+				if (myds->mypolls==NULL) {
+					thread->mypolls.add(POLLIN|POLLOUT, mybe->server_myds->fd, mybe->server_myds, thread->curtime);
+				}
+				int rc=myconn->async_change_user(myds->revents);
+				if (rc==0) {
+					myds->myconn->userinfo->set(client_myds->myconn->userinfo);
+					st=previous_status.top();
+					previous_status.pop();
+					NEXT_IMMEDIATE(st);
+				} else {
+					if (rc==-1) {
+						// the command failed
+						int myerr=mysql_errno(myconn->mysql);
+						if (myerr > 2000) {
```

Comment:
Agreed! Yet this code needs to be improved, as right now it only differentiates between client's errors (communication to the server failed) and server's errors (the server answered with an error).
Would be a big improvement to have error handling based on type of error.
For example, a "Too many connections" should be treated differently than "Access denied" even if they are both server's errors